### PR TITLE
[#7254] Implemented new API endpoint: rc_get_library_features (main)

### PR DESCRIPTION
**Pull request by korydraughn**
This API endpoint does not modify any server state.

Running the unit test for this endpoint is enough to confirm whether it's good for merging.

### CODE REVIEW

**Comment by trel** [comment]
commit message has "the the"

**Comment by trel** [comment]
so... is this a new source of truth?

**Comment by korydraughn** [comment]
No. It is supposed to match the contents of library_features.h.

**Comment by trel** [comment]
so it can drift from library_features.h?    can't we define/respond based on what's defined over there?

**Comment by korydraughn** [comment]
We'd have to build similar machinery to the irods error map stuff to achieve that. Which is certainly an option.

**Comment by trel** [comment]
Yeah, not feeling good about a list that we know will continue to grow having to be kept in sync.

File: server/api/src/rs_get_library_features.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+#define IRODS_FEATURE(F)        {#F, IRODS_FEATURE_VALUE(F)}, // NOLINT(cppcoreguidelines-macro-usage)
+
+    // clang-format off
+    *_features = strdup(nlohmann::json{
+        IRODS_FEATURE(IRODS_HAS_LIBRARY_TICKET_ADMINISTRATION)
+        IRODS_FEATURE(IRODS_HAS_LIBRARY_ZONE_ADMINISTRATION)
+        IRODS_FEATURE(IRODS_HAS_API_ENDPOINT_SWITCH_USER)
+        IRODS_FEATURE(IRODS_HAS_LIBRARY_SYSTEM_ERROR)
+        IRODS_FEATURE(IRODS_HAS_FEATURE_PROXY_USER_SUPPORT_FOR_CLIENT_CONNECTION_LIBRARIES)
+        IRODS_FEATURE(IRODS_HAS_API_ENDPOINT_CHECK_AUTH_CREDENTIALS)
```
Comment (korydraughn) [inline server/api/src/rs_get_library_features.cpp:29]: I really wish this could be generated at compile-time.

Is it possible? Hmm ...

**Comment by korydraughn** [comment]
> Yeah, not feeling good about a list that we know will continue to grow having to be kept in sync.

This may not be possible with just C++. We may need an alternative solution to achieve what you're asking.

**Comment by korydraughn** [comment]
One way to handle this is to parse the library_features.h header file and produce something the API endpoint can consume.

That would give us an opportunity to embed a compile-time JSON string into the API implementation. Of course that means the build system has to launch the parser for us.

Perhaps this is an opportunity to learn the clang-tools for parsing the C++ language.

**Comment by korydraughn** [comment]
Created issue #7276.

**Comment by korydraughn** [comment]
The clang-tidy issues are harmless and cannot be fixed easily at this time.

Will add the pound if there are no other concerns/comments.

File: lib/api/include/irods/get_library_features.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+struct RcComm;
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/// Returns a JSON string containing information about the connected server's library features.
+///
+/// \param[in]  _comm     A pointer to a RcComm.
+/// \param[out] _features A pointer to a pointer that will hold the results.
```
Comment (alanking) [inline lib/api/include/irods/get_library_features.h:15]: Worth describing the structure of the JSON string?

**Review by alanking — COMMENTED** [review]
Sorry, didn't realize this was out of draft.

I think this is fine, but just wondering aloud whether this could be part of a new and improved `rcGetMiscSvrInfo`. Just an idea, not attached to it.

**Comment by korydraughn** [comment]
> I think this is fine, but just wondering aloud whether this could be part of a new and improved `rcGetMiscSvrInfo`. Just an idea, not attached to it.

It could. My only concern is the JSON output of this new endpoint. It has the potential to grow very large over time.

Overall, the idea doesn't sound bad. Consider capturing it in a new issue and we can circle back to it after 4.3.1 is released.

File: lib/api/include/irods/get_library_features.h (hunk 2 above)
Comment (korydraughn) [reply lib/api/include/irods/get_library_features.h]: Done.

**Comment by alanking** [comment]
Okay, opened an issue here: https://github.com/irods/irods/issues/7278

**Comment by korydraughn** [comment]
Cool. Any thoughts on the updated documentation?

**Review by alanking — APPROVED** [review]
Look good to me. Please squash and add # when happy with it

**Comment by korydraughn** [comment]
Added #

**Review by alanking — APPROVED** [review]

**Pull request merged**